### ISSUE TITLE
optim-reduce-fullsequence-cloning

### ISSUE THREAD

**Issue by aminere** [eternagame/EternaJS#552]
Avoid calling `this._sequence.slice(0)` while iterating over bases

**Comment by luxaritas** [comment]
Keep in mind there are a couple places this is done because the sequence gets mutated, but the "caller" (whoever is passing the sequence) expects it not to be mutated. For example the sequence in an UndoBlock or Solution should never change, but Pose2D has a sequence that can be mutated internally, so it must be copied in order to preserve the correct behavior. (This has led to a class of bugs in the past)

**Comment by everyday847** [comment]
Yes, this is deliberate sometimes. It's a defensive choice, so you will
have to extensively debug each place where you elect to make a change!

**Comment by aminere** [comment]
Roger that @luxaritas  and @everyday847 !
I think this should be safe as it only gets rid of cloning while iterating over the sequence
Here is the PR: https://github.com/eternagame/EternaJS/pull/553
If anything let me know!

**Comment by luxaritas** [comment]
That change definitely looks sensible - if we're actually slicing that many times that certainly looks quite expensive!
